File: tools/pattern_generator.js
#!/usr/bin/env node

/**
 * ASCII Art Pattern Generator (Node.js)
 * Generate various ASCII art patterns with customizable parameters.
 */

class PatternGenerator {
  /**
   * Generate a spiral pattern
   */
  static spiral(size = 10, char = '●') {
    const gridSize = size * 2 + 1;
    const grid = Array(gridSize).fill().map(() => Array(gridSize).fill(' '));
    const centerX = size, centerY = size;
    
    let angle = 0;
    let radius = 0;
    const maxRadius = size;
    
    while (radius < maxRadius) {
      const x = Math.floor(centerX + radius * Math.cos(angle));
      const y = Math.floor(centerY + radius * Math.sin(angle));
      if (x >= 0 && x < gridSize && y >= 0 && y < gridSize) {
        grid[y][x] = char;
      }
      angle += 0.3;
      radius += 0.1;
    }
    
    return grid.map(row => row.join('')).join('\n');
  }

  /**
   * Generate a mandala pattern
   */
  static mandala(rings = 5, points = 8) {
    const size = rings * 4 + 1;
    const grid = Array(size).fill().map(() => Array(size).fill(' '));
    const center = Math.floor(size / 2);
    
    for (let ring = 1; ring <= rings; ring++) {
      const radius = ring * 2;
      for (let i = 0; i < points * ring; i++) {
        const angle = 2 * Math.PI * i / (points * ring);
        const x = Math.floor(center + radius * Math.cos(angle));
        const y = Math.floor(center + radius * Math.sin(angle));
        if (x >= 0 && x < size && y >= 0 && y < size) {
          grid[y][x] = ring % 2 === 0 ? '●' : '○';
        }
      }
    }
    
    grid[center][center] = '◉';
    return grid.map(row => row.join('')).join('\n');
  }

  /**
   * Generate a wave pattern
   */
  static wave(amplitude = 5, frequency = 0.5, width = 60) {
    const height = amplitude * 2 + 1;
    const lines = [];
    
    for (let y = 0; y < height; y++) {
      let line = '';
      for (let x = 0; x < width; x++) {
        const waveY = amplitude + amplitude * Math.sin(frequency * x);
        if (Math.abs(y - waveY) < 0.5) {
          line += '~';
        } else {
          line += ' ';
        }
      }
      lines.push(line);
    }
    
    return lines.join('\n');
  }

  /**
   * Generate a grid pattern
   */
  static grid(rows = 10, cols = 10, style = 'box') {
    const chars = style === 'box' ? {
      h: '─', v: '│', cross: '┼',
      tl: '┌', tr: '┐', bl: '└', br: '┘',
      tDown: '┬', tUp: '┴', tRight: '├', tLeft: '┤'
    } : {
      h: '·', v: '·', cross: '·',
      tl: '·', tr: '·', bl: '·', br: '·',
      tDown: '·', tUp: '·', tRight: '·', tLeft: '·'
    };
    
    const lines = [];
    const cellWidth = 3;
    
    // Top border
    let line = chars.tl;
    for (let col = 0; col < cols; col++) {
      line += chars.h.repeat(cellWidth);
      line += col < cols - 1 ? chars.tDown : chars.tr;
    }
    lines.push(line);
    
    // Rows
    for (let row = 0; row < rows; row++) {
      // Cell row
      line = chars.v;
      for (let col = 0; col < cols; col++) {
        line += ' '.repeat(cellWidth) + chars.v;
      }
      lines.push(line);
      
      // Separator
      if (row < rows - 1) {
        line = chars.tRight;
        for (let col = 0; col < cols; col++) {
          line += chars.h.repeat(cellWidth);
          line += col < cols - 1 ? chars.cross : chars.tLeft;
        }
        lines.push(line);
      }
    }
    
    // Bottom border
    line = chars.bl;
    for (let col = 0; col < cols; col++) {
      line += chars.h.repeat(cellWidth);
      line += col < cols - 1 ? chars.tUp : chars.br;
    }
    lines.push(line);
    
    return lines.join('\n');
  }

  /**
   * Generate a diamond pattern
   */
  static diamond(size = 7) {
    const lines = [];
    
    // Top half
    for (let i = 0; i < size; i++) {
      const spaces = ' '.repeat(size - i - 1);
      const stars = '◆'.repeat(2 * i + 1);
      lines.push(spaces + stars);
    }
    
    // Bottom half
    for (let i = size - 2; i >= 0; i--) {
      const spaces = ' '.repeat(size - i - 1);
      const stars = '◆'.repeat(2 * i + 1);
      lines.push(spaces + stars);
    }
    
    return lines.join('\n');
  }

  /**
   * Generate a circular pattern
   */
  static circle(radius = 10, char = '○') {
    const size = radius * 2 + 1;
    const grid = Array(size).fill().map(() => Array(size).fill(' '));
    const center = radius;
    
    for (let y = 0; y < size; y++) {
      for (let x = 0; x < size; x++) {
        const distance = Math.sqrt((x - center) ** 2 + (y - center) ** 2);
        if (Math.abs(distance - radius) < 0.7) {
          grid[y][x] = char;
        }
      }
    }
    
    return grid.map(row => row.join('')).join('\n');
  }

  /**
   * Generate a labyrinth pattern
   */
  static labyrinth(size = 9) {
    const grid = Array(size).fill().map(() => Array(size).fill('█'));
    
    // Simple maze generation
    function carve(x, y) {
      grid[y][x] = ' ';
      const dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]];
      
      // Shuffle directions
      for (let i = dirs.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [dirs[i], dirs[j]] = [dirs[j], dirs[i]];
      }
      
      for (const [dx, dy] of dirs) {
        const nx = x + dx * 2;
        const ny = y + dy * 2;
        if (nx > 0 && nx < size - 1 && ny > 0 && ny < size - 1 && grid[ny][nx] === '█') {
          grid[y + dy][x + dx] = ' ';
          carve(nx, ny);
        }
      }
    }
    
    carve(1, 1);
    return grid.map(row => row.join('')).join('\n');
  }

  /**
   * Generate a zigzag pattern
   */
  static zigzag(width = 40, height = 10) {
    const lines = [];
    for (let y = 0; y < height; y++) {
      let line = '';
      for (let x = 0; x < width; x++) {
        const pattern = Math.floor(x / 3) % 2;
        if ((y % 4 === 0 && pattern === 0) || (y % 4 === 2 && pattern === 1)) {
          line += '╱';
        } else if ((y % 4 === 0 && pattern === 1) || (y % 4 === 2 && pattern === 0)) {
          line += '╲';
        } else {
          line += ' ';
        }
      }
      lines.push(line);
    }
    return lines.join('\n');
  }
}

// CLI Interface
function main() {
  const args = process.argv.slice(2);
  
  if (args.length === 0) {
    console.log('ASCII Art Pattern Generator');
    console.log('\nUsage: node pattern_generator.js <pattern> [options]');
    console.log('\nPatterns:');
    console.log('  spiral [size] [char]        - Generate spiral (default: 10, ●)');
    console.log('  mandala [rings] [points]    - Generate mandala (default: 5, 8)');
    console.log('  wave [amp] [freq] [width]   - Generate wave (default: 5, 0.5, 60)');
    console.log('  grid [rows] [cols] [style]  - Generate grid (default: 10, 10, box)');
    console.log('  diamond [size]              - Generate diamond (default: 7)');
    console.log('  circle [radius] [char]      - Generate circle (default: 10, ○)');
    console.log('  labyrinth [size]            - Generate labyrinth (default: 9)');
    console.log('  zigzag [width] [height]     - Generate zigzag (default: 40, 10)');
    console.log('\nExamples:');
    console.log('  node pattern_generator.js spiral 15 ★');
    console.log('  node pattern_generator.js mandala 7 12');
    console.log('  node pattern_generator.js grid 5 5 dots');
    return;
  }

  const pattern = args[0];

  try {
    switch (pattern) {
      case 'spiral':
        console.log(PatternGenerator.spiral(
          parseInt(args[1]) || 10,
          args[2] || '●'
        ));
        break;
      case 'mandala':
        console.log(PatternGenerator.mandala(
          parseInt(args[1]) || 5,
          parseInt(args[2]) || 8
        ));
        break;
      case 'wave':
        console.log(PatternGenerator.wave(
          parseInt(args[1]) || 5,
          parseFloat(args[2]) || 0.5,
          parseInt(args[3]) || 60
        ));
        break;
      case 'grid':
        console.log(PatternGenerator.grid(
          parseInt(args[1]) || 10,
          parseInt(args[2]) || 10,
          args[3] || 'box'
        ));
        break;
      case 'diamond':
        console.log(PatternGenerator.diamond(parseInt(args[1]) || 7));
        break;
      case 'circle':
        console.log(PatternGenerator.circle(
          parseInt(args[1]) || 10,
          args[2] || '○'
        ));
        break;
      case 'labyrinth':
        console.log(PatternGenerator.labyrinth(parseInt(args[1]) || 9));
        break;
      case 'zigzag':
        console.log(PatternGenerator.zigzag(
          parseInt(args[1]) || 40,
          parseInt(args[2]) || 10
        ));
        break;
      default:
        console.error(`Unknown pattern: ${pattern}`);
        process.exit(1);
    }
  } catch (error) {
    console.error('Error generating pattern:', error.message);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

module.exports = PatternGenerator;

File: tools/pattern_generator.py
#!/usr/bin/env python3
"""
ASCII Art Pattern Generator
Generate various ASCII art patterns with customizable parameters.
"""

import math
import argparse
from typing import List, Tuple


class PatternGenerator:
    """Generate various ASCII art patterns."""
    
    @staticmethod
    def spiral(size: int = 10, char: str = "●") -> str:
        """Generate a spiral pattern."""
        grid = [[' ' for _ in range(size * 2 + 1)] for _ in range(size * 2 + 1)]
        center_x, center_y = size, size
        
        angle = 0
        radius = 0
        max_radius = size
        
        while radius < max_radius:
            x = int(center_x + radius * math.cos(angle))
            y = int(center_y + radius * math.sin(angle))
            if 0 <= x < len(grid[0]) and 0 <= y < len(grid):
                grid[y][x] = char
            angle += 0.3
            radius += 0.1
        
        return '\n'.join([''.join(row) for row in grid])
    
    @staticmethod
    def fractal_tree(depth: int = 5, branch_angle: float = 30) -> str:
        """Generate a fractal tree pattern."""
        lines = []
        
        def draw_branch(x: float, y: float, angle: float, length: float, level: int):
            if level == 0 or length < 1:
                return
            
            # Calculate end point
            end_x = x + length * math.cos(math.radians(angle))
            end_y = y - length * math.sin(math.radians(angle))
            
            # Store line segment
            lines.append((int(x), int(y), int(end_x), int(end_y)))
            
            # Recursive branches
            new_length = length * 0.7
            draw_branch(end_x, end_y, angle + branch_angle, new_length, level - 1)
            draw_branch(end_x, end_y, angle - branch_angle, new_length, level - 1)
        
        # Start from bottom center
        height = 40
        width = 80
        draw_branch(width // 2, height, 90, 10, depth)
        
        # Create grid
        grid = [[' ' for _ in range(width)] for _ in range(height)]
        
        for x1, y1, x2, y2 in lines:
            if 0 <= x1 < width and 0 <= y1 < height:
                grid[y1][x1] = '│' if abs(x2 - x1) < abs(y2 - y1) else '/'
        
        return '\n'.join([''.join(row) for row in grid])
    
    @staticmethod
    def mandala(rings: int = 5, points: int = 8) -> str:
        """Generate a mandala pattern."""
        size = rings * 4 + 1
        grid = [[' ' for _ in range(size)] for _ in range(size)]
        center = size // 2
        
        for ring in range(1, rings + 1):
            radius = ring * 2
            for i in range(points * ring):
                angle = 2 * math.pi * i / (points * ring)
                x = int(center + radius * math.cos(angle))
                y = int(center + radius * math.sin(angle))
                if 0 <= x < size and 0 <= y < size:
                    if ring % 2 == 0:
                        grid[y][x] = '●'
                    else:
                        grid[y][x] = '○'
        
        # Add center
        grid[center][center] = '◉'
        
        return '\n'.join([''.join(row) for row in grid])
    
    @staticmethod
    def wave(amplitude: int = 5, frequency: float = 0.5, width: int = 60) -> str:
        """Generate a wave pattern."""
        lines = []
        height = amplitude * 2 + 1
        
        for y in range(height):
            line = []
            for x in range(width):
                wave_y = amplitude + amplitude * math.sin(frequency * x)
                if abs(y - wave_y) < 0.5:
                    line.append('~')
                else:
                    line.append(' ')
            lines.append(''.join(line))
        
        return '\n'.join(lines)
    
    @staticmethod
    def grid_pattern(rows: int = 10, cols: int = 10, style: str = "box") -> str:
        """Generate a grid pattern."""
        if style == "box":
            h_line = "─"
            v_line = "│"
            cross = "┼"
            corner_tl = "┌"
            corner_tr = "┐"
            corner_bl = "└"
            corner_br = "┘"
            t_down = "┬"
            t_up = "┴"
            t_right = "├"
            t_left = "┤"
        else:  # dots
            h_line = "·"
            v_line = "·"
            cross = "·"
            corner_tl = corner_tr = corner_bl = corner_br = "·"
            t_down = t_up = t_right = t_left = "·"
        
        lines = []
        cell_width = 3
        
        # Top border
        line = corner_tl
        for col in range(cols):
            line += h_line * cell_width
            line += t_down if col < cols - 1 else corner_tr
        lines.append(line)
        
        # Rows
        for row in range(rows):
            # Cell row
            line = v_line
            for col in range(cols):
                line += " " * cell_width
                line += v_line
            lines.append(line)
            
            # Separator (except after last row)
            if row < rows - 1:
                line = t_right
                for col in range(cols):
                    line += h_line * cell_width
                    line += cross if col < cols - 1 else t_left
                lines.append(line)
        
        # Bottom border
        line = corner_bl
        for col in range(cols):
            line += h_line * cell_width
            line += t_up if col < cols - 1 else corner_br
        lines.append(line)
        
        return '\n'.join(lines)
    
    @staticmethod
    def sierpinski(depth: int = 4) -> str:
        """Generate Sierpinski triangle."""
        size = 2 ** depth
        grid = [[' ' for _ in range(size * 2)] for _ in range(size)]
        
        def draw_triangle(x: int, y: int, size: int):
            if size == 1:
                grid[y][x] = '▲'
                return
            
            half = size // 2
            draw_triangle(x, y, half)  # Top
            draw_triangle(x - half, y + half, half)  # Bottom left
            draw_triangle(x + half, y + half, half)  # Bottom right
        
        draw_triangle(size, 0, size)
        return '\n'.join([''.join(row) for row in grid])
    
    @staticmethod
    def diamond(size: int = 7) -> str:
        """Generate a diamond pattern."""
        lines = []
        
        # Top half (including middle)
        for i in range(size):
            spaces = ' ' * (size - i - 1)
            stars = '◆' * (2 * i + 1)
            lines.append(spaces + stars)
        
        # Bottom half
        for i in range(size - 2, -1, -1):
            spaces = ' ' * (size - i - 1)
            stars = '◆' * (2 * i + 1)
            lines.append(spaces + stars)
        
        return '\n'.join(lines)


def main():
    parser = argparse.ArgumentParser(description='Generate ASCII art patterns')
    parser.add_argument('pattern', choices=[
        'spiral', 'tree', 'mandala', 'wave', 'grid', 'sierpinski', 'diamond'
    ], help='Pattern type to generate')
    parser.add_argument('--size', type=int, default=10, help='Pattern size')
    parser.add_argument('--depth', type=int, default=5, help='Recursion depth (for tree/sierpinski)')
    parser.add_argument('--rings', type=int, default=5, help='Number of rings (for mandala)')
    parser.add_argument('--points', type=int, default=8, help='Points per ring (for mandala)')
    parser.add_argument('--amplitude', type=int, default=5, help='Wave amplitude')
    parser.add_argument('--frequency', type=float, default=0.5, help='Wave frequency')
    parser.add_argument('--width', type=int, default=60, help='Pattern width')
    parser.add_argument('--rows', type=int, default=10, help='Grid rows')
    parser.add_argument('--cols', type=int, default=10, help='Grid columns')
    parser.add_argument('--style', default='box', choices=['box', 'dots'], help='Grid style')
    parser.add_argument('--char', default='●', help='Character to use')
    
    args = parser.parse_args()
    
    generator = PatternGenerator()
    
    if args.pattern == 'spiral':
        print(generator.spiral(args.size, args.char))
    elif args.pattern == 'tree':
        print(generator.fractal_tree(args.depth))
    elif args.pattern == 'mandala':
        print(generator.mandala(args.rings, args.points))
    elif args.pattern == 'wave':
        print(generator.wave(args.amplitude, args.frequency, args.width))
    elif args.pattern == 'grid':
        print(generator.grid_pattern(args.rows, args.cols, args.style))
    elif args.pattern == 'sierpinski':
        print(generator.sierpinski(args.depth))
    elif args.pattern == 'diamond':
        print(generator.diamond(args.size))


if __name__ == '__main__':
    main()
